resolwe/test/testcases/__init__.py

```diff
@@ -34,17 +34,132 @@
+            # test on another thread).
+            # Remove already created directories and try again.
+            for dir_ in created_dirs:
+                os.rmdir(dir_)
+            continue
+
+    return created_dirs
+
+
+@override_settings(FLOW_EXECUTOR=FLOW_EXECUTOR_SETTINGS)
+@override_settings(FLOW_DOCKER_MAPPINGS=FLOW_DOCKER_MAPPINGS)
+@override_settings(CELERY_ALWAYS_EAGER=True)
+class TestCaseHelpers(DjangoSimpleTestCase):
     """Mixin for test case helpers."""
 
+    def _pre_setup(self, *args, **kwargs):
+        # NOTE: This is a work-around for Django issue #10827
+        # (https://code.djangoproject.com/ticket/10827) that clears the
+        # ContentType cache before permissions are setup.
+        ContentType.objects.clear_cache()
+        super(TestCaseHelpers, self)._pre_setup(*args, **kwargs)
+
+    def _override_executor_settings(self):
+        """Override Resolwe Flow Executor's settings to avoid clashes.
+
+        When Resolwe's tests are run in parallel or from different
+        Resolwe's code bases on the same system (e.g. on a CI server),
+        we must ensure executor's test directories and containers have
+        unique names to avoid clashes.
+
+        """
+        flow_executor_settings = copy.copy(getattr(settings, 'FLOW_EXECUTOR', {}))
+
+        # Create unique executor's test directories.
+        executor_dirs = ('DATA_DIR', 'UPLOAD_DIR')
+        for exec_dir, new_dir_path in zip(
+                executor_dirs,
+                _test_dirs(*(flow_executor_settings[d] for d in executor_dirs))):
+            flow_executor_settings[exec_dir] = new_dir_path
+
+        # Update FLOW_DOCKER_MAPPINGS setting.
+        flow_docker_mappings = _get_updated_docker_mappings(flow_executor_settings)
+
+        # Create unique container name prefix.
+        flow_executor_settings['CONTAINER_NAME_PREFIX'] = '{}_{}_{}'.format(
+            flow_executor_settings.get('CONTAINER_NAME_PREFIX', 'resolwe'),
+            get_random_string(length=6),
+            os.path.basename(flow_executor_settings['DATA_DIR'])
+        )
+
+        self.settings = override_settings(FLOW_EXECUTOR=flow_executor_settings,
+                                          FLOW_DOCKER_MAPPINGS=flow_docker_mappings)
+
+    def _clean_up(self):
+        """Clean up after test."""
+        if not self._keep_data:
+            shutil.rmtree(settings.FLOW_EXECUTOR['DATA_DIR'], ignore_errors=True)
+            shutil.rmtree(settings.FLOW_EXECUTOR['UPLOAD_DIR'], ignore_errors=True)
+
+        self.settings.disable()
+
+    def setUp(self):
+        """Prepare environment for test."""
+        super(TestCaseHelpers, self).setUp()
+
+        self._override_executor_settings()
+        self.settings.enable()
+
+        self._keep_data = False
+
+        self.addCleanup(self._clean_up)
+
+    def keep_data(self, mock_purge=True):
+        """Do not delete output files after test."""
+        self._keep_data = True
```

Comment:
So this will be moved to the test runner later?

Reply:
Yes, that's the plan.